[FLINK-26121] Adds queue cleanup to `ZooKeeperLeaderRetrievalConnectionHandlingTest`

XComp opened the pull request:
## What is the purpose of the change

We experienced some unstable ZooKeeper tests. This test belongs to the same group. See further details on the investigation effort in FLINK-26121.

## Brief change log

* Refactors test to use utility method for common functionality:
  * Instantiating the `QueueLeaderElectionListener`
  * Closing the driver
 * Add a method for cleaning up the queue after the test passed

## Verifying this change

* Nothing added to test the behavior

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalConnectionHandlingTest.java:
```diff
@@ -291,55 +271,72 @@ private void writeLeaderInformationToZooKeeper(
+                            ZooKeeperLeaderRetrievalDriver, QueueLeaderElectionListener, Exception>
+                    testCallback)
+            throws Exception {
         final QueueLeaderElectionListener queueLeaderElectionListener =
```
Why not model this as an extension that can be reused more easily?

XComp replied:
Good point 👍

XComp commented:
Test failure unrelated: FLINK-26272

XComp commented:
Thanks for the approval, @zentol . I reorganized and squashed the commits into meaningful chunks and rebased the branch. I will merge the changes after the build succeeded.
